### PR TITLE
initial PR 

### PR DESCRIPTION
**Pull request by nicktate**
not 100% done, but can you give it a quick look over and see if anything stands out 

@normanjoyner

### CODE REVIEW

File: handlers/containership.js [hunk 1]
```diff
@@ -1,5 +1,48 @@
...
+
+        _.forEach(sortedVersions, (version) => {
+            if(semver.satisfies(req.params.version, plugin.versions[version].supportedContainershipVersions)) {
+                latestValidPluginVersion = version;
+                return false;
+            }
+        });
+
+        if (latestValidPluginVersion === null) {
+            return res.sendStatus(400);
```
Comment (normanjoyner) [inline handlers/containership.js]: does 404 make sense here?

File: handlers/containership.js [hunk 2]
```diff
@@ -1,5 +1,48 @@
...
+        }
+
+        return res.json({
+            name: plugin.name,
+            description: plugin.description,
+            version: latestValidPluginVersion
+        });
+    }).catch((err) => {
+        console.error(err);
```
Comment (normanjoyner) [inline handlers/containership.js]: implement a lightweight logger eventually

File: handlers/plugins.js [hunk 3]
```diff
@@ -1,25 +1,218 @@
...
+        return res.json(plugin);
+    }).catch((err) => {
+        console.error(err);
+        return res.sendStatus(500);
+    });
 }
 
-module.exports.create = function(req, res, next) {
-    return next();
+function create(isPutRequest) {
```
Comment (normanjoyner) [inline handlers/plugins.js]: i'd prefer an object keyed similar to variable name here, only for clarity during invocation, but if it's only ever going to be this simple (no additional params) i'm cool with this

File: handlers/plugins.js [hunk 4]
```diff
@@ -1,25 +1,228 @@
...
+const redis = require('../lib/redis');
+
+const _ = require('lodash');
+const bluebird  = require('bluebird');
+const Joi = require('joi');
+
+const validateAsync = bluebird.promisifyAll(Joi.validate);
+
+const pluginSchema = Joi.object().keys({
+    name: Joi.string(),
```
Comment (normanjoyner) [inline handlers/plugins.js]: do we need `.required()` on all of these?

File: handlers/plugins.js [hunk 5]
```diff
@@ -1,25 +1,228 @@
...
+         *      "types": ["core", "cli"],
+         *      "supportedContainershipVersions": "> 1.8.0"
+         *   }
+         * }
+         */
+    )
+});
+
+const versionSchema = Joi.object().keys({
+    types: Joi.array().items(Joi.string().valid('core', 'cli')),
```
Comment (normanjoyner) [inline handlers/plugins.js]: same here

File: handlers/plugins.js [hunk 6]
```diff
@@ -1,25 +1,228 @@
...
+            }
+
+            plugin.versions[req.params.version] = req.body;
+            return redis.setAsync(req.params.plugin, JSON.stringify(plugin));
+        }).then((result) => {
+            if (result !== 'OK') {
+                return res.sendStatus(500);
+            }
+
+            return res.status(201).json(plugin);
```
Comment (normanjoyner) [inline handlers/plugins.js]: `201` only on POST

File: handlers/plugins.js [hunk 7]
```diff
@@ -1,25 +1,228 @@
...
+                return chain.break();
+            }
+
+            return redis.setAsync(req.params.plugin, JSON.stringify(plugin));
+        }).then((result) => {
+            if (result !== 'OK') {
+                return res.sendStatus(500);
+            }
+
+            return res.status(201).json(plugin);
```
Comment (normanjoyner) [inline handlers/plugins.js]: `201` only on POST

File: handlers/plugins.js [hunk 8]
```diff
@@ -1,25 +1,228 @@
...
-    return next();
+    const chain = redis.existsAsync(req.params.plugin).then((exists) => {
+        if (!exists) {
+            res.sendStatus(404);
+            return chain.break();
+        }
+
+        return redis.delAsync(req.params.plugin);
+    }).then((success) => {
+        return res.status(success === 1 ? 200 : 500).end();
```
Comment (normanjoyner) [inline handlers/plugins.js]: `204` if we're not sending anything back

File: handlers/plugins.js [hunk 9]
```diff
@@ -1,25 +1,228 @@
...
+        }
+
+        delete plugin.versions[req.params.version];
+        return redis.setAsync(req.params.plugin, JSON.stringify(plugin));
+    }).then((result) => {
+        if (result !== 'OK') {
+            return res.sendStatus(500);
+        }
+
+        return res.status(result === 'OK' ? 200 : 500).end();
```
Comment (normanjoyner) [inline handlers/plugins.js]: `204` if we're not sending anything back

File: handlers/plugins.js (hunk 3 above)
Comment (nicktate) [inline handlers/plugins.js]: should only ever be isPutRequest, just so I didn't have to dup all logic in handlers except for like two branches depending on whether it is a POST or PUT

**Pull request closed without merging**